wgsl/index.bs

```diff
@@ -4557,11 +4555,271 @@ TODO: *Stub*: Expression evaluation
+
+If the set of behaviors for a statement is a singleton, we take that to mean that reconvergence happened, and we pick the vertex corresponding to control-flow at the start of the statement as the vertex corresponding to control-flow at the exit of the statement, overriding the other rules below.
 
-### Divergence and reconvergence TODO ### {#divergence-reconvergence}
+The rules for analyzing statements take as argument both the statement itself and the vertex corresponding to control-flow at the beginning of it (which we'll note "CF" below) and return the following:
 
-### Uniformity restrictions TODO ### {#uniformity-restrictions}
+* A set of behaviors for the statement
+* A vertex corresponding to control-flow at the exit of it
+* A set of new vertices and edges to add to the graph
+
+In the table below, `CF1, S -> CF2, B` means "run the analysis on S starting with control-flow CF1, apply the required changes to the graph, and name the resulting control-flow CF2, and the resulting set of behaviors B".
+Similarly, `CF, E -> CF, V, B` means "run the analysis on expression E, starting with control-flow CF, apply the required changes to the graph, and name the resulting control-flow vertex CF, the resulting value vertex V, and the resulting set of behaviors B" (see next section for the analysis of expressions).
+
+We have a similar set of rules for expressions in left-value positions, that we denote by `CF, E => CF, L, B`. Instead of computing the vertex which corresponds to the uniformity of the value, it computes the vertex which corresponds to the uniformity of the memory location we are adressing.
+
+<table class='data'>
+  <caption>Uniformity rules for statements</caption>
+  <thead>
+    <tr><th>Statement<th>New vertices<th>Precondition<th>Behaviors<th>Resulting control-flow vertex<th>New edges
+  </thead>
+  <tr><td class="nowrap">*s1*; *s2*
+      <td>
+      <td class = "nowrap">*CF*, *s1* -> *CF1*, *B1*<br>
+          *CF1*, *s1* -> *CF2*, *B2*
+      <td>*B1* \ {Nothing} U *B2*
+      <td>*CF2*
+      <td>
+  <tr><td class="nowrap">if (*e*) then *s1* else *s2*
+      <td>*CFinside*, *CFend*
+      <td class="nowrap">*CF*, *e* -> *CF'*, *V*, *B*<br>
+          *CFinside*, *s1* -> *CF1*, *B1*<br>
+          *CFinside*, *s2* -> *CF2*, *B2*
+      <td>*B* U *B1* U *B2*
+      <td>*CFend*
+      <td class="nowrap">*CFinside* -> {*CF'*, *V*}<br>
+          *CFend* -> {*CF1*, *CF2*}
+  <tr><td class="nowrap">loop {*s1* continuing {*s2*}}
+      <td>*CF'*
+      <td class="nowrap">*CF'*, *s1* -> *CF1*, *B1*<br>
+          *CF1*, *s2* -> *CF2*, *B2*
+      <td>*B1* \ {Break, Continue} U *B2* U {Nothing}
+      <td>*CF'*
+      <td class="nowrap">*CF'* -> *CF2*
+  <tr><td class="nowrap">loop {*s1*}
+      <td>*CF'*
+      <td class="nowrap">*CF'*, *s1* -> *CF1*, *B1*
+      <td>*B1* \ {Break, Continue} U {Nothing}
```

Comment:
Are you sure that `loop` should `U {Nothing}`? Control flow will always enter the loop, and it can't exit without a `break`.

Reply:
You are correct, I took that part from the rules for while, which can exit whenever the condition is true.
The more precise behaviors here should be B1, removing Continue, and replacing Break by Nothing if it exists.
I'll make the change in the next commit.